src/libraries/System.Net.Security/src/System/Net/Security/Pal.Android/SafeDeleteSslContext.cs

```diff
@@ -75,61 +88,121 @@ protected override void Dispose(bool disposing)
+                    context._outputBuffer.Commit(dataLength);
+                }
+            }
+            catch (Exception ex)
+            {
+                Debug.Write("Exception Caught. - " + ex);
+                throw;
+            }
         }
 
         [UnmanagedCallersOnly]
         private static unsafe PAL_SSLStreamStatus ReadFromConnection(IntPtr connection, byte* data, int* dataLength)
         {
-            SafeDeleteSslContext? context = (SafeDeleteSslContext?)GCHandle.FromIntPtr(connection).Target;
-            Debug.Assert(context != null);
+            GCHandle h = GCHandle.FromIntPtr(connection);
+            Debug.Assert(h.IsAllocated);
 
-            int toRead = *dataLength;
-            if (toRead == 0)
-                return PAL_SSLStreamStatus.OK;
+            SafeDeleteSslContext? context = (SafeDeleteSslContext?)h.Target;
+            Debug.Assert(context != null);
 
-            if (context._inputBuffer.ActiveLength == 0)
+            try
             {
-                *dataLength = 0;
-                return PAL_SSLStreamStatus.NeedData;
+                lock (context._lock)
+                {
+                    if (context._disposed)
+                    {
+                        Debug.Write("ReadFromConnection: context is disposed");
+                        *dataLength = 0;
+                        return PAL_SSLStreamStatus.Error;
+                    }
+
+                    int toRead = *dataLength;
+                    if (toRead == 0)
+                        return PAL_SSLStreamStatus.OK;
+
+                    if (context._inputBuffer.ActiveLength == 0)
+                    {
+                        *dataLength = 0;
+                        return PAL_SSLStreamStatus.NeedData;
+                    }
+
+                    toRead = Math.Min(toRead, context._inputBuffer.ActiveLength);
+
+                    context._inputBuffer.ActiveSpan.Slice(0, toRead).CopyTo(new Span<byte>(data, toRead));
+                    context._inputBuffer.Discard(toRead);
+
+                    *dataLength = toRead;
+                    return PAL_SSLStreamStatus.OK;
+                }
             }
+            catch (Exception ex)
+            {
+                Debug.Write("Exception Caught. - " + ex);
+                throw;
+            }
+        }
 
-            toRead = Math.Min(toRead, context._inputBuffer.ActiveLength);
-
-            context._inputBuffer.ActiveSpan.Slice(0, toRead).CopyTo(new Span<byte>(data, toRead));
-            context._inputBuffer.Discard(toRead);
-
-            *dataLength = toRead;
-            return PAL_SSLStreamStatus.OK;
+        [UnmanagedCallersOnly]
+        private static void CleanupManagedContext(IntPtr managedContextHandle)
+        {
+            if (managedContextHandle != IntPtr.Zero)
+            {
+                GCHandle handle = GCHandle.FromIntPtr(managedContextHandle);
+                if (handle.IsAllocated)
+                {
```

Comment:
```suggestion
```
This is unnecessary. Non-null `managedContextHandle` is always going to turn into Allocated GCHandle (even if the handle was freed somewhere else already).

GCHandle is struct that is as dangerous as unsafe code. If you call Free on one copy of the struct, the other copies will still report IsAllocated==true even through the handle was freed.